Wallet fixes

jagerman opened the pull request:
Fixes some issues preventing the wallet from working properly on current stagenet:

- refreshing was broken because the get blocks binary endpoint was doing the wrong thing with no-miner-tx blocks.
- sending a transaction failed on stagenet because the wallet requests hard fork info for v16, but stagenet skips v16.  This is changed so then such a request now returns the starting height of the next highest hardfork (and returns an ending height of that value minus 1, so that the range is empty, but we still have a useful start point).

Doy-lee commented on src/cryptonote_basic/hardfork.cpp:
```diff
@@ -231,6 +231,14 @@ std::pair<std::optional<uint64_t>, std::optional<uint64_t>> get_hard_fork_height
+                // where the rules begin, and .second to one less (so that the range is empty, but
+                // we still return the .first value where the request HF rules become active).
+                found.first = hf.height;
+                found.second = hf.height > 0 ? hf.height - 1 : 0;
```
Find this behaviour a little bit unusual of returning `hf.height - 1` for the end of the hardfork. I only found one place where this value was used in:

```cpp 
void core_rpc_server::invoke(HARD_FORK_INFO& hfinfo, rpc_context) {
    const auto& blockchain = m_core.blockchain;
    auto version = hfinfo.request.version > 0 ? static_cast<hf>(hfinfo.request.version)
                 : hfinfo.request.height > 0 ? blockchain.get_network_version(hfinfo.request.height)
                                             : blockchain.get_network_version();
    hfinfo.response["version"] = version;
    hfinfo.response["enabled"] = blockchain.get_network_version() >= version;
    auto heights = get_hard_fork_heights(m_core.get_nettype(), version);
    if (heights.first)
        hfinfo.response["earliest_height"] = *heights.first;
    if (heights.second)
        hfinfo.response["latest_height"] = *heights.second;
    hfinfo.response["status"] = STATUS_OK;
}
```

Corresponding docs for the RPC call is 

```
/// Outputs:
...
///   the blockchain height is at or above the requested hardfork).
/// - `earliest_height` -- Block height at which the hard fork will become enabled.
/// - `last_height` -- The last block height at which this hard fork will be active; will be
///   omitted if this oxend is not aware of any following hard fork.
```

I'm ok with preserving the `nullopt` behaviour- or alternatively changing it to `UINT64_MAX` which would remove one edge case when dealing with the RPC response (e.g. you don't need to  check for field presence and you can always do `last_height - earliest_height` to calculate distance to next hard fork since the invariant `last_height >= earliest_height` is maintained).

jagerman replied:
I've reverted the change to `get_hard_fork_heights`, and instead addressed it by:

- Fix db_sqlite code that was using `get_hard_fork_heights` instead of `hard_fork_begins()` to look for when HF-dependent features become active.

- Improve hardfork.h documentation about the above issue.

- Make a similar change to the get_hard_forks RPC endpoint to return the results for the first hard fork >= the requested one instead of the exactly requested version.
  - This fixes the wallet on stagenet which makes requests to this endpoint for stagenet-skipped heights (e.g. 16) for testing feature availability.

jagerman committed:
Fix GET_BLOCKS_BIN handling of no-miner-tx transactions

Wallet refreshing was broken by this.

jagerman committed:
Improve handling of skipped hard forks

- Fix db_sqlite code that was using `get_hard_fork_heights` instead of
  `hard_fork_begins()` to look for when HF-dependent features become
  active.

- Improve hardfork.h documentation about the above issue.

- Make a similar change to the get_hard_forks RPC endpoint to return the
  results for the first hard fork >= the requested one.
  - This fixes the wallet on stagenet which makes requests to this
    endpoint for stagenet-skipped heights (e.g. 16) for testing feature
    availability.
